server/sonar-server/src/main/java/org/sonar/server/rule/RegisterRules.java

```diff
@@ -572,4 +576,39 @@ private void update(DbSession session, RuleDefinitionDto rule) {
     dbClient.ruleDao().update(session, rule);
   }
 
+
+  private static void verifyRuleKeyConsistency(List<RulesDefinition.ExtendedRepository> repositories) {
+    List<RulesDefinition.Rule> allRules = repositories.stream()
+      .flatMap(r -> r.rules().stream())
+      .collect(toList());
+
+    Set<RuleKey> definedRuleKeys = allRules.stream()
+      .map(r -> RuleKey.of(r.repository().key(), r.key()))
+      .collect(toSet());
+
+    List<RuleKey> definedDeprecatedRuleKeys = allRules.stream()
+      .flatMap(r -> r.deprecatedRuleKeys().stream())
+      .collect(toList());
+
+    Set<RuleKey> duplicates = findDuplicates(definedDeprecatedRuleKeys);
+    checkState(duplicates.isEmpty(), "The following deprecated rule keys are declared at least twice [%s]",
+      duplicates.stream().map(RuleKey::toString).collect(Collectors.joining(",")));
+
+    Sets.SetView<RuleKey> intersection = intersection(new HashSet<>(definedRuleKeys), new HashSet<>(definedDeprecatedRuleKeys));
+    checkState(intersection.isEmpty(), "The following rule keys are declared both as deprecated and used key [%s]", intersection.stream()
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Invoke method(s) only conditionally. ([squid:S2629](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2629))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Feh%2FSONAR-10310&issues=AWFmszu1ZGaLdh2cMSuQ&open=AWFmszu1ZGaLdh2cMSuQ)

[ISSUE_KEY]: AWFmszu1ZGaLdh2cMSuQ